py/kubeflow/testing/run_e2e_workflow.py

```diff
@@ -62,7 +71,14 @@ def get_namespace(args):
     return "kubeflow-releasing"
   return "kubeflow-test-infra"
 
-class WorkflowComponent(object):
+# imports py_func
+def py_func_import(py_func, kwargs):
+  path, module = py_func.rsplit('.', 1)
+  mod = importlib.import_module(path)
+  met = getattr(mod, module)
+  return met(**kwargs)
+
+class WorkflowKSComponent(object):
```

Comment:
Can we simplify this code by having a single class:

````
class WorkflowComponent(object):
   def  __init__(self, root_dir, data):
       ...
```

Where data is the dictionary loaded from the YAML file.

It doesn't look like there is much difference in the handling of PyWorkflows and KSWorkflows.
It looks like the only difference is how we initialize certain values; e.g. app_dir.

It makes sense to move this processing into the init function.

Using a single class will help eliminate code duplication which should avoid problems caused by slightly different implementations.

You can then define appropriate helper methods e.g.

```
  @property 
  def include_dirs:
     return self.data.get("include_dirs", [])
```

This would simplify the code below because you can remove the has_attr("include_dirs") checks and just do w.include_dirs

Reply:
If we do simplify both classes into a single one I imagine it would look something like this (I think instead of creating the helper methods we can just use default values), do you agree?:

```
class WorkflowComponent(object):
  """Datastructure to represent either a ksonnet component or Python function to
  submit a workflow."""
  def __init__(self, name, app_dir, component, job_types = [], include_dirs = [], params = {}, py_func, kwargs = {}):
    self.name = name
    self.app_dir = app_dir
    self.component = component
    self.job_types = job_types
    self.include_dirs = include_dirs
    self.params = params
    self.py_func = py_func
    self.args = kwargs
```

Reply:
The data parameter is the dict that is loaded from the yaml. The result of loading the yaml is a dictionary with field workflows which is a list of dictionaries so we can just do

```
config = yaml.load(<path to file>)

workflows = []
for w in config["workflows"]
  workflows.append(WorkflowComponent(w))
```
So I think it makes more sense to do the parsing of that dict inside the init function rather than outside the dict.

This way we only have to write the parsing logic once even if we were to construct workflow objects from the YAML in multiple places.

Also IIRC using objects as default values is not a good idea in python because python passes everything by reference. So each instance of WorkflowComponent would end up having self.job_types pointing to the same list. Which is not what we want.

Reply:
If we pass in the dict then we could just do

```
def __init__(self, data):
   self.job_types = data.get("job_types", [])
```